### DISCUSSION TITLE
Avoid defining `sizet_trait` and `type_trait` for every structure

### DISCUSSION THREAD

**Discussion by weslleyspereira** [tlapack/tlapack#51] [Ideas]
We can deduce the `size_type<matrix_t>` from the return of the function `nrows( matrix_t )` and deduce `type_t<matrix_t>` from `matrix_t()(0,0)`

```c++
    template< class matrix_t >
    using type_t = std::decay< decltype( std::declval<matrix_t>()(0,0) ) >::type;
    
    template< class matrix_t >
    using size_type = std::decay< decltype( std::declval<matrix_t>().nrows() ) >::type;
```

Two benefits:

- Reduce the number of definitions on the adapters for matrices and vectors.

- More compatibility among types. "Size type will be the type from the actual size value"

This can be extended to accept both matrices and vectors.

**Comment by thijssteel** [comment]
good idea

**Comment by weslleyspereira** [comment]
This was implemented in #78
